### DISCUSSION TITLE
[A-L] (18 giugno 2024) Esercizio 3

### DISCUSSION THREAD

**Discussion by FeddyLix17** [sapienzastudentsnetwork/progettazione-di-algoritmi#93] [Esami A-L]
Dare lo pseudocodice di un algoritmo che prende in input un grafo non-diretto e connesso con ogni vertice colorato o rosso o blu e restituisce in output un albero di copertura con un numero massimo di archi con termine dello stesso colore (quindi il numero massimo di archi colorati rosso-rosso oppure blu-blu).

**Comment by cornflexxx** [comment]
Per ogni $v\in V(G)$ cerchiamo ogni adiacente $u$ tale che $C(u)= C(v)$ e lo mettiamo in $\mathcal{S}$, in questo modo ci assicuriamo di aggiungere tutti gli archi con stesso vertice (eccetto quelli che creano un ciclo). Quindi il funzionamento è lo stesso di un qualsiasi algoritmo per creare un albero di copertura.

```
Esercizio3(G):
 S = {}
 for v in G:
      for u in v.adiacenti:
                if(C(u) = C(v)):
                      S.add((u,v))
 R = E(G)-S
 ST = {}
 for (u,v) in S:
       if ST + {(u,v)}.findcycle() == False:
               ST.add((u,v))
 for (u,v) in R:
          if ST + {(u,v)}.findcycle() == False:
               ST.add((u,v))

 return ST

```

**Comment by Elia-Belli** [comment]
## Greedy (Simile a Kruskal)
Volendo favorire la presenza di archi che terminano con lo stesso colore, ordiniamo $E(G)$ in modo da porre tali archi prima di quelli con vertici di colori differenti.
Nella costruzione dell'albero di copertura selezioniamo un arco se esso non introduce cicli, e al termine siccome $G$ è connesso saranno stati coperti tutti i nodi.
- **Dimostrazione lasciata al lettore**
```python
def albero_copertura_razista(G,C):                     #O(n+m)

	Order = []
	for (u,v) in G.E:                              #O(m)
		if(C[u] = C[v]): Same.push((u,v))
		else: Same.append((u,v))

	Sol = {}
	for (u,v) in Order:                            #O(m)
		if(no_cycles(Sol + {(u,v)})):          #O(n)
			Sol.add((u,v))
			
	return Sol
```

**Reply by daromacarlo** [reply]
razzista!?
Ti piacciono i razzi?

**Comment by daromacarlo** [comment]
```c
/*
Esame 18 Giugno 2024 - esercizio 3 #93
Dare lo pseudocodice di un algoritmo che prende in input un grafo non-diretto e connesso con ogni vertice
colorato o rosso o blu e restituisce in output un albero di copertura con un numero massimo 
di archi con termine dello stesso colore (quindi il numero massimo di archi colorati rosso-rosso oppure blu-blu).
*/


//algoritmo greedy
//MODIFICO L'ALGORITMO DI PRIM
def albero_colorato(G){
        sol = []
	vis = [0,...,0]
	while exists w in V(G) | vis[w] == 0{
		if exists (v,u) = arco in E(G) | (v.colore == u.colore && vis[u] == 0 && vis[v] == 1){
			sol.append(v,u)
		}
		else{
			(v,u) = arco in E(G)| (vis[u] == 0 && vis[v] == 1)
			sol.append(v,u)
		}
	}
}
return sol

/* 
costo finale = O(nm) 
   dato che il ciclo while è ripetuto m volte mentre le operazioni interne al ciclo costano O(2m).
*/
```
Domandone: all'esame vuole la dimostrazione della correttezza?

**Comment by Near047** [comment]
Io spero vada bene, faccio prim ma prim(a) cambio i pesi degli archi in modo che se il colore è uguale il peso dell'arco sarà -1 altrimenti 1. Il peso sarà O(m) + O(nm) = O(nm)

(Oppure invece di usare Prim, posso semplicemente usare Kruskal per essere più sintattico e non dover scegliere un vertice)

` Function prim_n(G):

    for ogni vertice e ∈ E(G) do
     (x,y) = e 
            if color(x) == color(y) then
                w(e) = -1
            else
                w(e) = +1
            end
        end
    end
    scegli un vertice v ∈ V(G)
    Sol = ∅
    R = {v}
    while R ≠ V(G) do
        (x, y) = arco con peso minimo tale che
                 x ∈ R e y ∈ V(G) − R
        Sol.add((x, y))
        R.add(y)
    end
    return Sol

 end `
